Add NamedTupleVariate and ProductNamedTupleDistribution

sethaxen opened the pull request:
Implements #1762

sethaxen commented:
One challenge is that `eltype` for a distribution `d` doesn't tell us what type `rand(d)` will return. The worst case here is `LKJCholesky`, for which the eltype is `Float64`, but `rand` returns a `Cholesky{Float64}`. So without calling `rand` recursively on all distributions once, we cannot know the exact types a call to `rand(::ProductNamedTupleDistribution)` should return.

devmotion commented:
> One challenge is that eltype for a distribution d doesn't tell us what type rand(d) will return. The worst case here is LKJCholesky, for which the eltype is Float64, but rand returns a Cholesky{Float64}. So without calling rand recursively on all distributions once, we cannot know the exact types a call to rand(::ProductNamedTupleDistribution) should return.

IMO `eltype` is a design flaw and should be removed completely. So I don't think it should block the PR.

sethaxen commented:
> IMO `eltype` is a design flaw and should be removed completely. So I don't think it should block the PR.

Without `eltype`, how would you recommend container eltype be determined for `rand` calls such as `rand(d, 2, 3)`? So far I've assumed this is what `eltype` should return, but then of course there's `LKJCholesky`, which returns `Float64` when the container eltype is `Cholesky{Float64}`.

devmotion commented:
No, that's not what `eltype` is supposed to give you currently. It is intended to return the element type of a single variate (e.g., `Float64` for a `MvNormal` whose variates are `Vector{Float64}`). But IMO there are at least three major problems: 1) It's implemented inconsistently and everyone expects something different; 2) It's a misnomer (IMO it was a suboptimal decision in retrospect to re-use `eltype` in Random; the less known `randtype` seems much clearer); 3) The concept breaks down if you try to allow variates of dynamic types (depending on the return type of the `rand(rng)` calls and/or the type of the parameters). In my experience any heuristic for addressing 3) is doomed to fail in some cases, so my suggestion would be to call `rand` if you want to know the type of its return value (and if you're adventurous you could try to infer it with `return_type` but IMO that's too brittle and internal for a default definition in Distributions).

sethaxen commented:
> the less known `randtype` seems much clearer

In which package is `randtype` defined? I have not heard of it.

> It is intended to return the element type of a single variate (e.g., `Float64` for a `MvNormal` whose variates are `Vector{Float64}`).

Ah, okay, then this PR currently implements it incorrectly. Will fix.

> my suggestion would be to call `rand` if you want to know the type of its return value

Cool, this is the approach I took, so will leave as is.

devmotion commented:
> In which package is randtype defined? I have not heard of it.

Sorry, I misremembered, it's called `gentype`. It's defined in Random, it also appeared in some discussion in some issue in Distributions, but IIRC there were PRs that tried to remove it, so not everyone shares my opinion 😄.

sethaxen commented:
I added a new API function `marginal`. While we have `component` for mixture distributions, we didn't have any API function for accessing marginals of a product distribution. And there are other cases where we know the marginals and users might want them (e.g. slices of `MvNormal`). It's an optional addition, and I'm fine with removing if maintainers think this should be in its own PR.

RE docs, currently `product_distribution` is documented in the multivariate distributions page, but with this PR, sometimes `product_distribution` will return a different variate form, so I think it makes sense to move all docs related to product distributions to their own page. `marginal` (if kept) could be documented here as well, though it's more general than that.

sethaxen commented:
@devmotion this is ready for a review

Red-Portal commented:
Hi all, would it be possible to get this moving? I think it would be really great to have this feature!

sethaxen commented:
This is really just waiting on a review. @devmotion?

Red-Portal commented:
Hi @devmotion, could we make this happen?

devmotion commented on src/common.jl:
```diff
@@ -16,6 +16,12 @@ const Univariate    = ArrayLikeVariate{0}
+`F <: NamedTupleVariate{K}` specifies that the variate or a sample is of type
+`NamedTuple{K}`.
+"""
+abstract type NamedTupleVariate{K} <: VariateForm end
```
Couldn't/Shouldn't this be a concrete subtype?
```suggestion
struct NamedTupleVariate{K} <: VariateForm end
```

sethaxen replied:
Perhaps, but this would be inconsistent with other `VariateForm` subtypes. Both `ArrayLikeVariate` and `CholeskyVariate` could be concrete subtypes but are currently defined as abstract subtypes.

devmotion replied:
I see. I'm not sure there's a good reason for the current implementation but maybe let's be consistent for now then.

devmotion replied:
I thought a bit more about this and I think the existing definitions might just be a bug - I've never encountered a subtype of these variate forms, and I wouldn't expect anyone to define a subtype for NamedTupleVariate either.

I also wonder if abstract type parameters might actually be worse for the compiler.

So I would suggest making it a concrete type.

devmotion commented on src/common.jl:
```diff
@@ -464,6 +470,16 @@ Base.@propagate_inbounds function loglikelihood(
     return sum(Base.Fix1(logpdf, d), x)
 end
 
+"""
+    marginal(d::Distribution, k...) -> Distribution
+
+Return the marginal distribution of `d` at the indices `k...`.
+
+The result is the distribution of the variate `rand(d)[k...]` that one would obtain by
+integrating over all other indices.
+"""
+marginal(d::Distribution, k...)
+
```
Maybe this should go into a separate PR? It seems this is a bit orthogonal to NamedTupleVariate.

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,131 @@
+end
+function ProductNamedTupleDistribution(
+    dists::NamedTuple{K,V}
+) where {K,V<:Tuple{Vararg{Distribution}}}
```
```suggestion
) where {K,V<:Tuple{Distribution,Vararg{Distribution}}}
```
to ensure that there is at least one distribution

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,131 @@
+
+function _rand!(rng::AbstractRNG, d::ProductNamedTupleDistribution, xs::AbstractArray)
+    for i in eachindex(xs)
+        xs[i] = Random.rand(rng, d)
```
```suggestion
        xs[i] = rand(rng, d)
```

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,131 @@
+function Base.rand(rng::AbstractRNG, d::ProductNamedTupleDistribution{K}) where {K}
+    return NamedTuple{K}(map(Base.Fix1(rand, rng), d.dists))
+end
+function Base.rand(rng::AbstractRNG, d::ProductNamedTupleDistribution, dims::Dims)
+    xs = return map(CartesianIndices(dims)) do _
+        return rand(rng, d)
+    end
```
I think for efficiency reasons you might want to use something like
```suggestion
sampler(d::ProductNamedTupleDistribution{K}) where K = ProducNamedTupleSampler{K}(map(sampler, d.dists))
function Base.rand(rng::AbstractRNG, d::ProductNamedTupleDistribution, dims::Dims)
    spl = sampler(d)
    xs = return map(CartesianIndices(dims)) do _
        return rand(rng, spl)
    end
```
Since `sampler` can be type unstable, probably it would be good to introduce a function barrier though.

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,131 @@
+    return xs
+end
+
+function _rand!(rng::AbstractRNG, d::ProductNamedTupleDistribution, xs::AbstractArray)
```
Also here, probably you should use `sampler` + a function barrier.

sethaxen commented:
Bump @devmotion

devmotion commented:
> what do you think about this proposed docs restructure?

I think it's reasonable to move it.

sethaxen commented:
@devmotion I restructured the docs. I also noticed that, while `Product` is deprecated, its docstring was missing a deprecation warning, so I added it. I also figured the behavior of `mean(::ProductNamedTuple)`, `mode`, `var`, etc may not be obvious, so I added a doctested example.

sethaxen commented:
Bump @devmotion

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,152 @@
+function kldivergence(
+    d1::ProductNamedTupleDistribution{K}, d2::ProductNamedTupleDistribution{K}
+) where {K}
+    return mapreduce(kldivergence, +, d1.dists, d2.dists)
```
Maybe be consistent with the definitions above and use

```suggestion
    return sum(map(kldivergence, d1.dists, d2.dists))
```
?

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,152 @@
+_product_namedtuple_eltype(dists) = typejoin(map(_gentype, dists)...)
+
+function Base.show(io::IO, d::ProductNamedTupleDistribution)
+    return show_multline(io, d, collect(pairs(d.dists)))
```
Is this the standard definition in Distributions? I though `show` with two arguments should print a short one-line representation and only the three argument version should be more verbose.

sethaxen replied:
There's only one 3-arg definition in Distributions (for `Censored`). Otherwise it's always 2-arg. And a bunch of distributions override the default `show` method to explicitly call `show_multline`: `Truncated`, `MvNormal`, `MvLogNormal`, `Censored`, and all matrix distributions.

Without this override, `show_multline` gets used anyways but with the only parameter being the field `dists`.

devmotion replied:
I see, we don't want to touch this issue in this PR here 🙂

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,152 @@
+end
+_gentype(::Distribution) = Any
+
+_product_namedtuple_eltype(dists) = typejoin(map(_gentype, dists)...)
```
This is a bit inconsistent with `ProductDistribution` which uses `promote_type`, not `typejoin`. Maybe similarly one could use something like

```suggestion
_product_namedtuple_eltype(dists::NamedTuple{K,V}) where {K,V} = __product_promote_type(eltype, V)
```
?

sethaxen replied:
I think we want this inconsistency right?, since the eltype of the concatenation of 2 vectors uses promotion, while the eltype of the merging of 2 (named-)tuples uses a typejoin (ultimately calling https://github.com/JuliaLang/julia/blob/v1.10.5/base/tuple.jl#L228-L237). Shouldn't `eltype(::Distribution)` in general agree with `eltype(rand(::Distribution))`?

devmotion replied:
> Shouldn't eltype(::Distribution) in general agree with eltype(rand(::Distribution))?

I guess that's what distributions should satisfy currently (even though it's quite inconsistent). So I think you're right, we should use the same approach that is used for computing `eltype(::NamedTuple)`, ie. we should use `typejoin`.

sethaxen commented:
@devmotion alright, I've implemented the requested changes.

devmotion commented on test/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,195 @@
+
+    @testset "show" begin
+        d = ProductNamedTupleDistribution((x=Gamma(1.0, 2.0), y=Normal()))
+        @test sprint(show, d) == """
```
FYI an alternative would be
```suggestion
        @test repr(d) == """
```

devmotion reviewed (approved):
Looks good to me, thank you 🙂

sethaxen commented:
@devmotion can you merge this and make a release?

devmotion commented on src/namedtuple/productnamedtuple.jl:
```diff
@@ -0,0 +1,155 @@
+julia> nt = rand(d)
+(x = 1.5155385995160346, y = [0.533531876438439, 0.466468123561561])
+
+julia> pdf(d, nt)
```
Just today I was dealing with ordering of `NamedTuple` in TransformVariables transforms: https://github.com/tpapp/TransformVariables.jl/pull/120

So I wonder: Should we have
```julia
julia> pdf(d, (; x = 1.5155385995160346, y = [0.533531876438439, 0.466468123561561])) == pdf(d, (; y = [0.533531876438439, 0.466468123561561], x = 1.5155385995160346))
true
```
?

sethaxen replied:
Good point, I modified the code to support permutations of the names.
